new-packages/ts-project/__tests__/source-edits/add-action.test.ts

```diff
@@ -0,0 +1,543 @@
+    {
+      type: 'add_action',
+      path: ['a'],
+      actionPath: ['exit', 0],
+      name: 'doStuff',
+    },
+  );
+  expect(await project.applyTextEdits(textEdits)).toMatchInlineSnapshot(`
+    {
+      "index.ts": "import { createMachine } from "xstate";
+
+    createMachine({
+      initial: "a",
+      states: {
+        a: {
+          exit: "doStuff"
+        },
+      },
+    });",
+    }
+  `);
+});
+
+test('should be possible to add an entry action to an existing single entry action', async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        entry: "bark",
+      });
+    `,
+  });
+
+  const project = await createTestProject(tmpPath);
+
+  const textEdits = project.editDigraph(
+    {
+      fileName: 'index.ts',
+      machineIndex: 0,
+    },
+    {
+      type: 'add_action',
+      path: [],
+      actionPath: ['entry', 1],
+      name: 'doStuff',
+    },
+  );
+  expect(await project.applyTextEdits(textEdits)).toMatchInlineSnapshot(`
+    {
+      "index.ts": "import { createMachine } from "xstate";
+
+    createMachine({
+      entry: [
+        "bark",
+        "doStuff"
+      ],
+    });",
+    }
+  `);
+});
+
+test('should be possible to add an exit action to an existing single exit action', async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        exit: "bark",
+      });
+    `,
+  });
+
+  const project = await createTestProject(tmpPath);
+
+  const textEdits = project.editDigraph(
+    {
+      fileName: 'index.ts',
+      machineIndex: 0,
+    },
+    {
+      type: 'add_action',
+      path: [],
+      actionPath: ['exit', 1],
+      name: 'doStuff',
+    },
+  );
+  expect(await project.applyTextEdits(textEdits)).toMatchInlineSnapshot(`
+    {
+      "index.ts": "import { createMachine } from "xstate";
+
+    createMachine({
+      exit: [
+        "bark",
+        "doStuff"
+      ],
+    });",
+    }
+  `);
+});
+
+test('should be possible to add an entry action at the start of the existing entry actions', async () => {
+  const tmpPath = await testdir({
+    'tsconfig.json': JSON.stringify({}),
+    'index.ts': ts`
+      import { createMachine } from "xstate";
+
+      createMachine({
+        entry: ["bark", "meow"],
+      });
+    `,
+  });
+
+  const project = await createTestProject(tmpPath);
+
+  const textEdits = project.editDigraph(
+    {
+      fileName: 'index.ts',
+      machineIndex: 0,
+    },
+    {
+      type: 'add_action',
+      path: [],
+      actionPath: ['entry', 0],
+      name: 'doStuff',
+    },
+  );
+  expect(await project.applyTextEdits(textEdits)).toMatchInlineSnapshot(`
+    {
+      "index.ts": "import { createMachine } from "xstate";
+
+    createMachine({
+      entry: ["doStuff",
+      "bark", "meow"],
```

Comment:
Multi-item array wrapping could use some work - it's not incorrect right now, it just ain't pretty. For that reason, I don't intend to improve this rn

Reply:
That's fine